### ISSUE TITLE
Network plugin not sending data via specific interface

### ISSUE THREAD

**Issue by elieyal** [collectd/collectd#2818]
*  Version of collectd: 5.8.0-3
*  Operating system / distribution: CentOS 7.3.1611
*  On the other side there is an InfluxDB listening to UDP port 80

## Expected behavior

I have several servers that acts the same. all have:
eth0 is an internal NIC - not connected to the internet.
eth1 another internal NIC - not connected to the internet
eth2 is an external NIC - has access to the internet
lo

when the collectd start with no specific interface configured it selects the eth0. Although there is no connection to the internet, with tcpdump, I can see the packets being sent.
When redirecting the network plugin to a specific network interface and looking at the tcpdump I would see the packets being sent as well

## Actual behavior

In this case, I don't see ANY packets going out of the configured interface (or any of the interfaces)
I have tested with traceroute that there is a connection between the 2 servers (run traceroute on the collectd server and tcpdump to listen on the receiving interface where the influxdb is installed - I can see the packets arriving so it is not a FW or network issue). The packets are not even going out TO the interface.

## Steps to reproduce
*  step 1
Configure collectd.conf:

```
<Plugin network>
    <Server "influxdb" "80">
        Interface "eth2"
    </Server>
</Plugin>
```
* step 2
    Restart collectd

* step 3
    run tcpdump:
    tcpdump -i eth2 -p -n -s 1500 udp port 80

I have to say that I have another server with the same configuration (a copy with same configuration) and it works as expected. This is why this problem is a great mystery.

What I would like to have:

1. A bind-address capability instead of the interface name. it might resolve some issues
2. How can I compile Collectd myself with debug capabilities? I haven't found anything in the Readme's or on any website.

**Comment by rpv-tomsk** [comment]
Hi!

>I have tested with traceroute that there is a connection between the 2 servers (run traceroute on the collectd server and tcpdump to listen on the receiving interface where the influxdb is installed - I can see the packets arriving so it is not a FW or network issue). The packets are not even going out TO the interface.

If you have correct routing between servers, then all should work without `Interface` specified.

>    A bind-address capability instead of the interface name. it might resolve some issues

Patches are welcome!

>    How can I compile Collectd myself with debug capabilities? I haven't found anything in the Readme's or on any website.

`./configure --enable-debug`

**Comment by elieyal** [comment]
Ok, I managed to compile the source with debug and got the debug working.
With some network/system guys I managed to understand the issue with binding to the interface name instead of the IP address:
in some cases, using tcpdump, we managed to see that when using the interface name for binding, the OS does an ARP request. only one server that is controlled in my environment, the router will reply to the ARP requests:
listening on eth2, link-type EN10MB (Ethernet), capture size 262144 bytes
08:27:56.878593 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
08:27:56.878869 ARP, Reply ***.***.***.*** is-at f8:b1:56:05:69:33, length 46
08:28:01.872171 IP 10.9.26.19.51061 > ***.***.***.***.80: UDP, length 1286
08:28:01.876592 IP 10.9.26.19.51061 > ***.***.***.***.80: UDP, length 1346
08:28:01.924598 IP 10.9.26.19.51061 > ***.***.***.***.80: UDP, length 1323
*** I Masked the IPs with '*'

In other cases, when the server is sitting in other environments and the router there is NOT responding to ARP for IPs that are not in it's own subnets, there is no ARP reply and the packets are not being sent anywhere:
This if from this king of server:
listening on eth2, link-type EN10MB (Ethernet), capture size 262144 bytes
05:20:04.715439 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:05.716416 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:06.718415 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:08.716485 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:09.718416 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:10.720427 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:12.716434 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:13.718457 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28
05:20:14.720415 ARP, Request who-has ***.***.***.*** tell **.*.**.**, length 28

BTW, the same issue will happen when you ping an IP with the -I option and choose the interface name over the IP address. when using the interface name, ARP requests will be sent while when using the IP address it will not.

ping -I eth2 **.***.***.**
Vs.
ping -I **.***.*.** **.***.***.**

Now the question is if someone can add an option to bind IP address instead (or as another option) of bind to interface name?

It should be really easy as all the infrastructure is there and the configuration will be:
```
<Plugin network>
    <Server "influxdb" "80">
        bindaddr "**.***.***.***"
    </Server>
</Plugin>
```
Can you please help with this?

**Comment by rpv-tomsk** [comment]
Why don't you simply set routing? I think this should help.

**Comment by elieyal** [comment]
In my case I simply cant use routing on the server because the server is already routing other stuff through different NICs. Also due to really harsh security, the connection out is limited and closed off in other NICs but the one that is open. This is why I must use bind-address and not the interface name.

**Comment by elieyal** [comment]
Hi,
I have a patch ready for this to make a bind-address option. It was tested and works properly.
[collectd-bind-address.zip](https://github.com/collectd/collectd/files/2110220/collectd-bind-address.zip)
what is the process of this to be released as a release in a RPM?

**Comment by elieyal** [comment]
Hi,
I have committed the change to the network.c file (verifies and passed the 2 steps).
When will be the next release with this change?

**Comment by rpv-tomsk** [comment]
As you closed a previous issue and deleted the branch, you have to push another one and create new PR.

In commit message you can add lines like `Closes: #2818` or `Issue: #2818`, the commit will reference to this issue then.

**Comment by rpv-tomsk** [comment]
Closed by #2831
